src/localisation/test/ekf_node_rostest.cpp

```diff
@@ -0,0 +1,189 @@
+                        double speed,
+                        double ang_vel,
+                        double seconds_traveling,
+                        std::mt19937 e2) {
+        // Set local variables
+        int i                   = 0;
+        double measured_x       = x;
+        double measured_y       = y;
+        double measured_angle   = angle;
+        double measured_speed   = speed;
+        double measured_ang_vel = ang_vel;
+        std::normal_distribution<> noise(0, 5);
+        sensor_msgs::Imu test_imu;
+        nav_msgs::Odometry test_gps;
+        nav_msgs::Odometry test_encoder;
+
+        for (i = 0; i < seconds_traveling; i++) {
+            // save every generated value as real position
+            x += sin(angle) * speed;
+            y += cos(angle) * speed;
+            angle += ang_vel;
+            angle = EKF::defineAngleInBounds(angle);
+
+            // save every generated values as measured position
+            measured_x += sin(angle) * speed + noise(e2);  // add some noise
+            measured_y += cos(angle) * speed + noise(e2);  // add some noise
+            measured_angle += ang_vel + (noise(e2) / 400); // add some noise
+            measured_angle = EKF::defineAngleInBounds(measured_angle);
+
+            measured_speed += noise(e2);           // add some random noise
+            measured_ang_vel += (noise(e2) / 200); // add some random noise
+
+            // publish IMU
+            test_imu.angular_velocity.z = measured_ang_vel;
+            test_imu.orientation =
+            tf::createQuaternionMsgFromYaw(measured_angle);
+            test_imu_publisher.publish(test_imu);
+            ros::Rate loop_rate(10);
+            loop_rate.sleep();
+            ros::spinOnce();
+
+            // publish encoder
+            test_encoder.twist.twist.linear.x = measured_speed;
+            test_encoder_publisher.publish(test_encoder);
+            loop_rate.sleep();
+            ros::spinOnce();
+
+            // publish GPS
+            test_gps.pose.pose.position.x = measured_x;
+            test_gps.pose.pose.position.y = measured_y;
+            test_gps_publisher.publish(test_gps);
+            loop_rate.sleep();
+            ros::spinOnce();
+        }
+    }
+};
+
+TEST_F(EKFNodeTest, testEKFNode) {
+    // set up gaussian noise generator
+    // std::random_device rd;
+    std::mt19937 e2(99); // std::mt19937 e2(rd()); for a random seed
+    // Start at the engineering design centre (in UTM)
+    double x       = 481917;  // starting positoin in metres
+    double y       = 5456662; // starting position in metres
+    double angle   = 0;       // starting angle in radians
+    double speed   = 3;       // starting speed in metres per second
+    double ang_vel = 0;       // starting angular velocity in radians per second
+    double seconds_traveling = 50;
+
+    // all sections of the rostest assume a dt of 1 second remember to
+    // change dt in EKFNode.cpp to reflect that
```

Comment:
Hmm, is this still true? I think we're now using ROS time in the node. Since we've separated the EKF out into a separate class, the best option here might be to change this from a ROSTest into a regular GTest and test the EKF directly (as we can provide `dt`) (though we could leave these tests here as well as general integration tests).

Reply:
left in as a general intergration test